internal/pkg/earpointer/heap.go

```diff
@@ -0,0 +1,209 @@
+// or a synthetic element, possibly annotated with some kind of context
+// (call stack, allocation site etc).
+
+// Reference is the base interface for local, global, and synthetic references.
+type Reference interface {
+	fmt.Stringer
+
+	Type() types.Type // Return the data type of the underlying element
+	Value() ssa.Value // Return the underlying local or global
+}
+
+// Context represents the calling context of a reference.
+// Typically, it contains a stack of call instructions.
+type Context []ssa.Instruction
+
+// Local is a heap partition that represents all abstract
+// objects that a register of reference type can point to in a specific context.
+type Local struct {
+	context *Context
+	reg     ssa.Value
+}
+
+func (l Local) Type() types.Type {
+	return l.reg.Type()
+}
+
+func (l Local) Value() ssa.Value {
+	return l.reg
+}
+
+func (l Local) String() string {
+	reg := l.reg
+	if p := reg.Parent(); p != nil {
+		return l.context.String() + p.Name() + "." + reg.Name()
+	}
+	return l.context.String() + reg.Name()
+}
+
+// Global is a heap partition that represents all abstract
+// objects that a global of reference type can point to.
+type Global struct {
+	global *ssa.Global
+}
+
+func (g Global) Type() types.Type {
+	return g.global.Type()
+}
+
+func (g Global) Value() ssa.Value {
+	return g.global
+}
+
+func (g Global) String() string {
+	return g.global.String()
+}
+
+type SyntheticKind int
+
+const (
+	SyntheticValueOf SyntheticKind = iota
+	SyntheticField
+)
+
+// Synthetic is for references created internally with an operator
+// and an operand reference. It is for creating references not directly associated
+// with an IR element. For example,
+// (1) *r0 can be represented by using operator SyntheticValueOf (*);
+// (2) a unknown field can be represented by using operator SyntheticField.
+//     For example, when a struct is cloned, some fields may not appear in the IR,
+//     these "unused" fields can be synthesized and stored in the heap.
+type Synthetic struct {
+	kind SyntheticKind
+	ref  Reference
+}
+
+func (s Synthetic) Type() types.Type {
+	return s.ref.Type()
+}
+
+func (s Synthetic) Value() ssa.Value {
+	return s.ref.Value()
+}
+
+func (s Synthetic) String() string {
+	if s.kind == SyntheticValueOf {
+		return "*" + s.ref.String()
+	}
+	return s.ref.String() + "[.]"
+}
+
+// Field can be (1) a struct field linked to an IR field (Var);
+// (2) or one with a name but not linked to any IR element (irField = nil).
+type Field struct {
+	Name string
+	//lint:ignore U1000 ignore dead code for now
+	irField *types.Var
+}
+
+// Helper to get the pseudo-field that is used to denote looking up an
+// array/slice/map index. For example, consider a register x of type T[], where
+// T is a struct type. The contents of the array/slice, namely x[i], is
+// over-approximated as the heap partition pointed-to by a pseudo-field named
+// getIndexField().
+//lint:ignore U1000 ignore dead code for now
+func getIndexField() Field {
```

Comment:
Nit: I think this would be slightly clearer as a package-level global. Same for `DirectPointToField` below.